plugin/igus_plugin.cpp

```diff
@@ -0,0 +1,308 @@
+        ServoID::R_ANKLE_ROLL,     ServoID::L_ANKLE_ROLL,     ServoID::HEAD_YAW,        ServoID::HEAD_PITCH,
+    };
+    const std::array<JointID, 20> servo_id_to_joint = {
+        JointID::R_SHOULDER_PITCH, JointID::L_SHOULDER_PITCH, JointID::R_SHOULDER_ROLL, JointID::L_SHOULDER_ROLL,
+        JointID::R_ELBOW,          JointID::L_ELBOW,          JointID::R_HIP_YAW,       JointID::L_HIP_YAW,
+        JointID::R_HIP_ROLL,       JointID::L_HIP_ROLL,       JointID::R_HIP_PITCH,     JointID::L_HIP_PITCH,
+        JointID::R_KNEE,           JointID::L_KNEE,           JointID::R_ANKLE_PITCH,   JointID::L_ANKLE_PITCH,
+        JointID::R_ANKLE_ROLL,     JointID::L_ANKLE_ROLL,     JointID::HEAD_YAW,        JointID::HEAD_PITCH,
+    };
+
+public:
+    // Inherit constructors
+    using ModelPlugin::ModelPlugin;
+
+    template <typename T>
+    using Network = NUClear::dsl::word::Network<T>;
+
+    /**
+     * The load function is called by Gazebo when the plugin is inserted into simulation
+     * @param model A pointer to the model that this plugin is attached to
+     * @param sdf   A pointer to the plugin's SDF element
+     */
+    void Load(physics::ModelPtr model, sdf::ElementPtr sdf) {
+
+        // Safety check to see if the SDF file is attached correctly
+        if (model->GetJointCount() == 0) {
+            gzerr << "Invalid joint count, NUbots Igus plugin not loaded" << std::endl;
+            return;
+        }
+
+        // Just output a message for now
+        gzdbg << "Attaching an iGus plugin to model [" << model->GetName() << "]" << std::endl;
+
+        // Store the model pointer for convenience
+        model = model;
+
+        // Get the imu for this model
+        this->imu_sensor = std::dynamic_pointer_cast<sensors::ImuSensor>(
+            sensors::get_sensor(model->SensorScopedName("imu_sensor")[0]));
+
+        // Get the joint pointers
+        joints = model->GetJoints();
+
+        // Set up the update event
+        update_connection = event::Events::ConnectWorldUpdateBegin(std::bind(&NUbotsIgusPlugin::update_robot, this));
+
+        get_reactor().on<Network<message::motion::ServoTargets>>().then(
+            [this](const message::motion::ServoTargets& targets) {
+                // Store the incoming commands in the queue for the next simulation frame
+                std::lock_guard<std::mutex> lock(command_mutex);
+                for (const auto& c : targets.targets()) {
+                    // Convert ServoID to JointID
+                    message::motion::ServoTarget joint_target(c);
+                    joint_target.set_id(servo_id_to_joint[c.id()]);
+                    joint_target.set_position(c.position() - joint_offsets[c.id()]);
+                    command_queue.push_back(joint_target);
+                }
+            });
+
+        // Setup a P-controller, with a gain of 4.
+        pid = common::PID(40.0, 0.0, 0.0);
```

Comment:
These magic numbers worry me

Reply:
what worries me is that is says 4, not 40